spirv_glsl.cpp

```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
 	return expr;
 }
 
+string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type)
+{
+	if (flattened_buffer_blocks.count(base))
+	{
+		std::string expr;
+		flattened_access_chain(expr, base, indices, count, target_type, 0);
+
+		return expr;
+	}
+	else
+	{
+		return access_chain(base, indices, count, false);
+	}
+}
+
+void CompilerGLSL::flattened_access_chain(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	if (!target_type.array.empty())
+	{
+		SPIRV_CROSS_THROW("Access chains that result in an array can not be flattened");
+	}
+	else if (target_type.basetype == SPIRType::Struct)
+	{
+		flattened_access_chain_struct(expr, base, indices, count, target_type, offset);
+	}
+	else if (target_type.columns > 1)
+	{
+		flattened_access_chain_matrix(expr, base, indices, count, target_type, offset);
+	}
+	else
+	{
+		flattened_access_chain_vector_scalar(expr, base, indices, count, target_type, offset);
+	}
+}
+
+void CompilerGLSL::flattened_access_chain_struct(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (size_t i = 0; i < target_type.member_types.size(); ++i)
+	{
+		if (i != 0)
+			expr += ",";
+
+		const SPIRType &member_type = get<SPIRType>(target_type.member_types[i]);
+		uint32_t member_offset = type_struct_member_offset(target_type, uint32_t(i));
+
+		flattened_access_chain(expr, base, indices, count, member_type, offset + member_offset);
+	}
+
+	expr += ")";
+}
+
+void CompilerGLSL::flattened_access_chain_matrix(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (uint32_t i = 0; i < target_type.columns; ++i)
+	{
+		if (i != 0)
+			expr += ",";
+
+		flattened_access_chain_vector_scalar(expr, base, indices, count, target_type, offset + i * 16);
```

Comment:
There is a MatrixStride decoration for stride between columns. I think 16 would be correct here, but probably a good idea to not hardcode it.